test/unit/offset.js

```diff
@@ -568,4 +568,142 @@ QUnit.test( "iframe scrollTop/Left (see gh-1945)", function( assert ) {
 	}
 } );
 
+(function() {
+	var POSITION_VALUES = [ "static", "relative", "absolute", "fixed" ];
+
+	supportjQuery.each( POSITION_VALUES, function( i, docPosition ) {
+		supportjQuery.each( POSITION_VALUES, function( i, bodyPosition ) {
+			supportjQuery.each( [ "static", "absolute" ], function( i, markerPosition ) {
+
+				testIframe( "coordinates relative to document" +
+							" (html{position:" + docPosition + "} body{position:" + bodyPosition + "} div{position:" + markerPosition + "})",
+						"offset/rel-doc.html", function( assert, $, iframe, doc ) {
+					assert.expect( 256 );
+
+					var docElem = doc.documentElement, // `<html>` element
+						bodyElem = doc.body,
+						marker1 = doc.getElementById( "marker1" ),
+						marker2 = doc.getElementById( "marker2" ),
+						$marker1 = $( marker1 ),
+						$marker2 = $( marker2 ),
+						affectProps, markerTopLeft,
+
+						// Bit as Flag of box-properties
+						FLAG_DOC_MARGIN = 1,
+						FLAG_DOC_BORDER = 2,
+						FLAG_DOC_PADDING = 4,
+						FLAG_BODY_MARGIN = 8,
+						FLAG_BODY_BORDER = 16,
+						FLAG_BODY_PADDING = 32,
```

Comment:
> You said that `situations()` should be called in each case of that 16 cases. Is my interpretation correct?

Not exactly., I'm saying that `situations` and "CASE 1"/"CASE 2"/"CASE 3" should be removed in favor of a simpler approach. The `testIframe` callback shouldn't need to do _too_ much:
```js
// Assume initial conditions include html/document margin/border/padding:
// html { font-size: 10px; margin: 1em; border: 2em solid red; padding: 4em; }
// body { margin: 8em; border: 16em solid blue; padding: 32em; }

// …and a static element and an absolutely-positioned element:
// #static { position: static; }
// #absolute { position: absolute; top: 2em, left: 2em; }

// Establish document-relative origins for children of <body>
docElem.style.position = docPosition;
body.style.position = bodyPosition;
var bodyContentOrigin = 10 + 20 + 40 + 80 + 160 + 320;
var origin = 0 +
	(docPosition === "static" ? 0 : 10 + 20) +
	(bodyPosition === "static" ? 0 : 40 + 80 + 160);

// Check offsets
var absoluteOffset = $( "#absolute" ).offset();
assert.deepEqual(
	$( "#static" ).offset(),
	{ top: bodyContentOrigin, left: bodyContentOrigin },
	"offset of position:static element includes <html> and <body> box styles" );
assert.deepEqual( absoluteOffset, { top: origin + 20, left: origin + 20 },
	"offset of position:absolute element ignores box styles of position:static ancestors" );
$( "#absolute" ).offset( absoluteOffset );
assert.deepEqual( $( "#absolute" ).offset(), absoluteOffset, "offset() round-trips" );

// Reposition html and body, tracking origin adjustments given margin/border/padding
var originAdjust = 0;
$( docElem ).css( { top: "1.5em", left: "1.5em" } );
if ( docPosition !== "static" ) {
	originAdjust += 15 - (docPosition === "relative" ? 0 : 10);
}
$( body ).css( { top: "3em", left: "3em" } );
if ( bodyPosition === "fixed" || bodyPosition === "absolute" && docPosition === "static" ) {

	// html box styles no longer matter
	origin = 80 + 160;
	bodyContentOrigin = origin + 320;
	originAdjust = 30;
} else if ( bodyPosition !== "static" ) {
	originAdjust += 30 - (bodyPosition === "relative" ? 0 : 40);
}

// Recheck offsets
absoluteOffset = $( "#absolute" ).offset();
assert.deepEqual(
	$( "#static" ).offset(),
	{ top: bodyContentOrigin + originAdjust, left: bodyContentOrigin + originAdjust },
	"offset of position:static element respects ancestor positioning" );
assert.deepEqual(
	absoluteOffset,
	{ top: origin + originAdjust + 20, left: origin + originAdjust + 20 },
	"offset of position:absolute respects ancestor positioning" );
$( "#absolute" ).offset( absoluteOffset );
assert.deepEqual( $( "#absolute" ).offset(), absoluteOffset, "offset() still round-trips" );
```

And there could easily be ways to simplify even further.